quodlibet/util/matcher.py

```diff
@@ -0,0 +1,382 @@
+
+        Size of a_items and b_items can differ. If there are more a_items than
+        there are b_items, None is used if no match could be assigned to an a.
+        The returned list always has the size of a_items.
+
+        In terms of performance, it's preferable to supply the smaller list as a_items.
+        """
+        if not b_items:
+            return [None for _ in a_items]
+
+        self._b_items = b_items
+        b_size = len(b_items)
+
+        self._b_idx_to_a_match_data = [None for _ in range(b_size)]
+        self._a_idx_to_match_data = []
+
+        for a_idx, a in enumerate(a_items):
+            match_data = _MatchData(a_idx, a, b_size)
+            self._a_idx_to_match_data.append(match_data)
+
+            self._measure_similarity_to_find_best_b_match(match_data)
+
+            if self._is_below_minimum_similarity(match_data):
+                match_data.best_b_idx = None
+            else:
+                self._handle_conflicts_if_any(match_data)
+
+        result = [md.best_b_idx for md in self._a_idx_to_match_data]
+
+        if self.should_store_similarity_matrix:
+            matrix = [md.b_idx_to_similarity for md in self._a_idx_to_match_data]
+            self.similarity_matrix = matrix
+
+        # cleanup
+        del self._b_items
+        del self._b_idx_to_a_match_data
+        del self._a_idx_to_match_data
+
+        return result
+
+    def _is_below_minimum_similarity(self, match_data):
+        cont_attr_idx = match_data.continue_attr_index
+        similarity = match_data.best_b_similarity
+
+        # If this hasn't gone through every attribute and weight yet, we normalize the
+        # similarity, so it's in the range of [0, 1]. Without this, minimum_similarity
+        # would be somewhat unpredictable.
+        #
+        # For example, a b_item might be undefeatable with a similarity of 0.52. If we
+        # wouldn't normalize it, it would not pass a minimum of 0.6, even if it could
+        # easily achieve that with the remaining weights.
+        if cont_attr_idx < len(self._attr_with_weight):
+            cur_total_weight = 1 - self._weight_left[cont_attr_idx]
+            similarity /= cur_total_weight
+
+        return similarity < self.minimum_similarity_ratio
+
+    def _handle_conflicts_if_any(self, a1_match_data):
+        while True:
+            best_b_idx = a1_match_data.best_b_idx
+            if best_b_idx is None:
+                # a1 has no matches left or could not find a match
+                return
+
+            a2_match_data = self._b_idx_to_a_match_data[best_b_idx]
+            if a2_match_data is None:
+                self._b_idx_to_a_match_data[best_b_idx] = a1_match_data
+                return
+
+            # We have found a conflict and will now solve it.
+            a1_match_data = self._get_worse_match_data(a1_match_data, a2_match_data)
+
+    def _get_worse_match_data(self, a1_match_data, a2_match_data):
```

Comment:
Why... worse... than what? :confused:

Reply:
I added a docstring and changed the name to `_get_losing_match_data`. (returns the match data that lost the conflict)

Maybe that's more clear?